src/cpp/ext/otel/otel_plugin.h

```diff
@@ -25,16 +25,37 @@
 
 #include <memory>
 #include <string>
+#include <utility>
+#include <vector>
 
 #include "absl/container/flat_hash_set.h"
 #include "absl/strings/string_view.h"
 #include "opentelemetry/metrics/meter_provider.h"
 #include "opentelemetry/metrics/sync_instruments.h"
 #include "opentelemetry/nostd/shared_ptr.h"
 
+#include "src/core/lib/transport/metadata_batch.h"
+
 namespace grpc {
 namespace internal {
 
+class LabelsInjector {
+ public:
+  virtual ~LabelsInjector() {}
+  // Read the incoming initial metadata to get the set of labels to be added to
+  // metrics. (Includes the local labels.)
+  virtual std::vector<std::pair<std::string, std::string>> GetPeerLabels(
```

Comment:
> Do we expect more than one value for the same label key? If not, why use a vector of pairs instead of a map?
Since the allocation of memory for a vector is contiguous, the iteration over a vector is faster than a map.

> The local labels (both keys and values) should be fixed for the entire process, so those should be able to be allocated only once, not once per RPC. Can we expose them as absl::string_view instead of std::string?

There are two kinds of local labels, 1) those that we need to send over to the peer and 2) those that we need to record along with our metrics (since they are not present in MonitoredResource)
For 1, yes, these are saved as a `grpc_core::Slice` and that's how we add them to the out going initial metadata
For 2, since OTel C++ accepts std::string and not absl::string_view, doing this, won't help us.

> The peer label keys and values could probably be exposed as Slice, which would take a ref instead of making a copy.
The same issue, right now, we don't have a way of passing a non std::string string type to OTel C++. 

> Maybe instead of using either a vector or a map, we can return an iterable interface?
I think we can pass an iterable interface to OTel C++, but we would first need the ability to pass in string_view.

Reply:
> Since the allocation of memory for a vector is contiguous, the iteration over a vector is faster than a map.

I'm not sure how much that really matters, since the strings themselves may not be stored inline, so we could be reading from different locations in memory anyway.  I guess that if a `std::string` happens to be inlined, it might help, but certainly if we switch to `absl::string_view`, this would still be an issue.

> since OTel C++ accepts std::string and not absl::string_view, doing this, won't help us.

I argue that even if OT currently does not support this, it's better for us to have the right interface between two components in our own code, so that if OT fixes this later, we have less work to do.  In other words, even if we still need to construct a `std::string` here, I think it's better to do it in the `CallTracer` than to bake this assumption into the `LabelsInjector` API.

> I think we can pass an iterable interface to OTel C++, but we would first need the ability to pass in string_view.

Unless I'm missing something here (and please tell me if I am), I think that's actually an orthogonal issue.  Even if OT requires `std::string`, we can still construct the `std::string` as we return each element of the iterable interface.  Doing that would still avoid the allocation for the vector itself.

Reply:
> I'm not sure how much that really matters, since the strings themselves may not be stored inline, so we could be reading from different locations in memory anyway. I guess that if a std::string happens to be inlined, it might help, but certainly if we switch to absl::string_view, this would still be an issue.

The difference may not be much, but I don't see a reason to use a map given that a simple vector satisfies our purpose.

> I argue that even if OT currently does not support this, it's better for us to have the right interface between two components in our own code, so that if OT fixes this later, we have less work to do. In other words, even if we still need to construct a std::string here, I think it's better to do it in the CallTracer than to bake this assumption into the LabelsInjector API.

I was wrong. The OTel API actually does take in an `absl::string_view`. It's just that the SDK later converts it to a std::string. Irrespective, you are right, I shouldn't unnecessarily create a string when an `absl::string_view` does the job. I wonder why this wasn't working for me earlier.. maybe it's because they use their own bespoke `opentelemetry::nostd::string_view` instead of `absl::string_view`

> Unless I'm missing something here (and please tell me if I am), I think that's actually an orthogonal issue. Even if OT requires std::string, we can still construct the std::string as we return each element of the iterable interface. Doing that would still avoid the allocation for the vector itself.

Created an iterator. In the new logic, I use a std::string to store the labels that are read from the peer while we use absl::string_view for our local labels

Reply:
The changes to use `absl::string_view` look good!  (As a side note, I realized that my earlier comment about using `Slice` doesn't make sense, since we're not encoding any of the label values as a single string by itself; we're always encoding them into a protobuf first.)

However, I think the way you're using the iterable here isn't sufficient to actually solve the problem that I was concerned about.  It looks like you're using the iterable between the `CallTracer` and OT, but you're still using a vector between the `LabelsInjector` and the `CallTracer`, so this isn't actually helping us reduce the unnecessary overhead between the components in our own code.

What I had intended here was to use an iterable interface to return the data from `LabelsInjector` to `CallTracer`, to avoid the need for allocating the vector.  For example, in `GetPeerLabels()`, the iterable could hold a slice ref to the metadata value.  On its first iteration, it could deserialize the protobuf.  It can have a fixed list of labels that it will look at and track the current index into that last, so that it knows which element to return on each iteration.  This would avoid the need to allocate a vector, and it would avoid the need to allocate a `std::string` for each label value (it could be returned as a `string_view`).

Note that once you have such an iterable interface between `LabelsInjector` and `CallTracer`, it should be trivial to wrap that iterable in the OT iterable interface that you need to pass the data from `CallTracer` to OT.